oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,357 @@
+package org.heigit.ohsome.oshdb.util.osh;
+
+import static java.lang.String.format;
+
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.heigit.ohsome.oshdb.osh.OSHEntity;
+import org.heigit.ohsome.oshdb.osh.OSHNode;
+import org.heigit.ohsome.oshdb.osh.OSHRelation;
+import org.heigit.ohsome.oshdb.osh.OSHWay;
+import org.heigit.ohsome.oshdb.osm.OSMEntity;
+import org.heigit.ohsome.oshdb.osm.OSMMember;
+import org.heigit.ohsome.oshdb.osm.OSMNode;
+import org.heigit.ohsome.oshdb.osm.OSMRelation;
+import org.heigit.ohsome.oshdb.osm.OSMWay;
+import org.heigit.ohsome.oshdb.util.OSHDBTimestamp;
+
+public class OSHEntityTimeUtils {
+  private OSHEntityTimeUtils() {
+    throw new IllegalStateException("utility class");
+  }
+
+  /**
+   * Returns the changeset ids which correspond to modifications of this entity.
+   *
+   * <p>Used internally to group modifications by changeset.</p>
+   *
+   * @param osh the osh entity to work on
+   * @return a map between timestamps and changeset ids
+   */
+  public static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHEntity osh) {
+    switch (osh.getType()) {
+      case NODE:
+        return getChangesetTimestamps((OSHNode) osh);
+      case WAY:
+        return getChangesetTimestamps((OSHWay) osh);
+      case RELATION:
+        return getChangesetTimestamps((OSHRelation) osh);
+      default:
+        throw new UnsupportedOperationException(format(UNSUPPORTED_OSMTYPE_MESSAGE, osh.getType()));
```

Comment:
I also often see in such a case an IlegalStateException as well. I think we should just check what is common to throw in such a case.

- do we not support an unknown osmtype
- does the osh have an illegal state because of the unknow osmtype

both are valid views i guess

Reply:
yeah, I thought about that as well. I found [this thread](https://stackoverflow.com/a/49749857/1627467) on SO:

> `UnsupportedOperationException` should be used as the method is not supported at all while `IllegalStateException` should be used as the method is supported but that in the current state, it is not legal.

They give the example of the `Iterator.remove` method which can be "unsupported" for some kinds of iterators (throwing `UnsupportedOperationException`), but might also throw `IllegalStateException` when they run into a situation where the method is implemented, but it's not possible to remove any more items because there are no more to remove.

In the end it doesn't really matter since this line cannot be triggered by "normal" code anyway since there are only the above listed 3 `OSMType`s, and the default case is just added to make the Java (11) compiler happy. If we eventually upgrade our Java version (to ⪆ 14), we could get rid of the `default` clause here (see [JEP 361](https://openjdk.java.net/jeps/361#Exhaustiveness).